yellow-paper/docs/gas-and-fees/gas-metering-and-fees.md

```diff
@@ -0,0 +1,683 @@
+
+If app logic reverts due to OutOfL2Gas, `l2GasLeft` is set to zero. Based on the update logic above, this results in 
+
+```
+public_kernel_public_inputs.gas_limits.l2_gas_limit ==
+  public_kernel_public_inputs.end.gas_used.l2_gas_used
+  + public_kernel_public_inputs.end_non_revertible.gas_used.l2_gas_used
+```
+
+Thus, the user is charged the full `l2_gas_limit`.
+
+However, we drop the DA and L1 gas used associated with app logic, i.e.:
+```
+public_kernel_public_inputs.end.gas_used.da_gas_used := 0
+public_kernel_public_inputs.end.gas_used.l1_gas_used := 0
+```
+
+:::caution
+This is a departure from the existing yellow paper which suggests that all gas be consumed in the event of app logic OutOfGas revert.
+:::
+
+### Handled nested OutOfL2Gas exceptions
+
+Suppose an enqueued public call begins execution with `l2GasLeft` of 1000.
+
+Suppose it performs a few operations that reduce this to 900.
+
+Suppose it calls a nested public call, specifying that the nested call can only consume 500 l2 gas.
+
+Suppose the nested call consumes 495 l2 gas, then tries to consume 10 more.
+
+The nested call will throw an OutOfL2Gas exception.
+
+Suppose the user handles this exception.
+
+Execution continues in the parent call, with `l2GasLeft` set to 400.
+
+
+### OutOfDAGas reverts in App Logic
+
+If app logic reverts due to OutOfDAGas, the user is only charged for DA gas and L1 gas consumed in the non-revertible phase.
+
+This matches the behavior of OutOfL2Gas reverts in app logic, i.e.:
+
+```
+public_kernel_public_inputs.end.gas_used.da_gas_used := 0
+public_kernel_public_inputs.end.gas_used.l1_gas_used := 0
+```
+
+**However,** `l2GasLeft` is left unchanged: the user is only charged for l2 gas used up to the point of the revert.
+
+### Handled nested OutOfDAGas exceptions
+
+Suppose an enqueued public call begins execution with `daGasLeft` of 1000.
+
+Suppose it performs a few operations that reduce this to 900.
+
+Suppose it calls a nested public call, specifying that the nested call can only consume 500 DA gas.
+
+Suppose the nested call consumes 495 DA gas, then tries to consume 10 more.
+
+The nested call will throw an OutOfDAGas exception.
+
+Suppose the user handles this exception.
+
+Execution continues in the parent call, with `daGasLeft` set to 900, since the DA gas that correspond to the state updates in the nested call **no longer exist**.
+
+:::caution Question
+does this place too much complexity on the VM?
+is this too kind? e.g. consider the the 4844 spec:
+> The actual blob_fee as calculated via calc_blob_fee is deducted from the sender balance before transaction execution and burned, and is not refunded in case of transaction failure.
+:::
+
+### OutOfL1Gas reverts in App Logic
+
+Same as OutOfDAGas reverts in app logic.
+
+### Handled nested OutOfL1Gas exceptions
+
+Same as handled nested OutOfDAGas exceptions in app logic.
+
+### All other reverts in App Logic
+
+These follow the same logic as OutOfDAGas reverts in app logic.
```

Comment:
Presumably not for OutOfL2Gas? If a nested call uses all of it's e.g. 100 L2 gas, that amount is still added to the accumulated gas used of it's caller?